codeforamerica/vita-min
# Fyst 186/create charitable contributions page

## Link to pivotal/JIRA issue
https://codeforamerica.atlassian.net/browse/FYST-186

## Sequence that the PR adds

```mermaid
sequenceDiagram
    AZ321Index ->> AZ321New: User has no contributions
    AZ321New -->> AZ321New: Contribution invalid
    AZ321New -->> AZ321Index: Contribution valid
    AZ321Index -->> AZ321Edit: Edit a contribution
    AZ321Edit -->> AZ321Index: User submits valid contribution
    AZ321Edit -->> AZ321Edit: Contribution invalid
    AZ321Edit -->> NextPath: User marks 'no contribution'
    AZ321Index ->> NextPath: User proceeds in the flow
    AZ321New -->> NextPath: User marks 'no contribution'
```

## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

## What was done?
- Adds an AZ 321 Qualified Organization Charitable Contribution page
- Adds a new date accessible concern which can be used in conjunction with `cga_date_select` to write regular validation rules
  - I'd like special attention paid to this as it is metaprogramming, which can incur a steep cost if left unchecked. I don't think this bit is particularly egregious but still, cleverness should be met with skepticism.

## How to test?

### Automated Tests

- Controller specs to test flow
- Model tests to test the unique form_create context for this model
- Some date specific testing in the model as well
- Date Accessible meta programming test added

### Manual testing

- Progress the intake as usual through the AZ flow, continuing to `/en/questions/az-qualifying-organization-contributions/`
- Page should automatically redirect to `/en/questions/az-qualifying-organization-contributions/new` on first visit
- Click continue
- Note that the page does not proceed, instead showing a validation error
- Check 'no' and click continue again
- Note that the page progresses to the next step
- Click 'go back'
- Click 'yes'
- Note form opens
- Click 'continue'
- Note that all fields are considered invalid
- Fill out text values for all fields and set the date
- Note that the date is only for the current tax year
- Click continue
- Note that the amount fails because the value is non-numeric
- Create a numeric value
- Click 'continue'
- Note that the page is now the index page and your contribution has been saved.
- Click edit
- Note that the form displays the values just entered, the form is fully visible, and 'yes' is checked
- Change a value
- Click continue
- Note that the changes are reflected on the page
- Click 'Delete' and note the confirmation window
- Accept
- Note the flash marking 'contribution deleted'
- Proceed to fill in three more contributions
- Note that the 'add another' button is gone (due to there being a limit of three)

## Screenshots

![image](https://github.com/user-attachments/assets/cd1b40a4-6543-4720-a186-e72422899ab4)

![image](https://github.com/user-attachments/assets/1ae5ce68-e883-4e6d-a720-a1bb79208623)
